# Analysis code for *Molecular convergence of risk variants for congenital heart defects leveraging a regulatory map of the human fetal heart*
All the snakemake pipelines were designed to be run on an HPC platform, [Sherlock](https://www.sherlock.stanford.edu) specifically. The config files are located in \[dir\]/config
## envrionments
  ### environment for snakemake runs:
  envs/snakemake.yml 
  ### enivronemnt for running R scripts:
  envs/R.yml

## scRNA-seq alignment and background removal and Sample demultiplexing
  ### RNA_alignment_snakemake
  This directory contains code for aligning FASTQ files of the RNA portion of multiomic sequencing to a reference assembly using [Starsolo](https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md) and demultiplexing samples using [Souporcell](https://github.com/wheaton5/souporcell). The input sample table requires information about the location of the FASTQ files, the name of the library as appeared in the FASTQ filenames, and the number of samples mixed in the library. 
  ### CellBender_Snakemake 
  This directory contains code for removing ambient RNA (background) using CellBender. 
  
## scATAC-seq alignment
  ### scATAC_pipeline_lite
  This directory contains code for aligning FASTQ files of the ATAC portion of multiomic sequencing to a reference assembly. The code base was forked from https://github.com/austintwang/scATAC_pipeline_lite (commit: 76a4bc545318b5685eb9843e94000dde0d3394e7) with modifications for file and result organization. 

## scATAC-seq quality control
  ### multiomic_ATAC_QC_Snakemake
  The directory contains code for quality control the scATAC-seq portion of the multiome libraries. As shown in the config file, the default filters are minTSS >= 6, and minFrags >= 1000, unless otherwise specified in the sample sheet. Only cells considered to be valid by Starsolo were considered. Both "addDoubletScores" from [ArchR](https://www.archrproject.com/bookdown/inferring-scatac-seq-doublets-with-archr.html) and [Amulet](https://github.com/UcarLab/AMULET) were used to identify doublets. 

## scRNA-seq quality control and normalization
  ### RNA_QC_Snakemake
  The directory contains code for generating the necessary QC metrics of scRNA-seq using the outputs from [ambient RNA removal](#cellbender_snakemake). 
  ### Normalization_Snakemake
  The directory contains code for filtering cells based on the thresholds set by manually examing the QC metrics generated by the previous step. The filtered scRNA count matrices are then normalized and [Scublet](https://github.com/swolock/scrublet) is used to identify doublets. 
  
## Doublet removal
  ### Doublet_removal_Snakemake
  This directory contains code for rerunning Scrublet if specified in the 
## Clustering and cell type annotation
  ### evaluation_Snakemake
  ### clustering_Snakemake
  ### annotation_Snakemake

## Defining gene programs with consensus non-negative matrix factorization
  ### cNMF_run_and_analysis/Snakefile_cNMF
  ### cNMF_run_and_analysis/Snakefile_cNMF_analysis

## Identifying transcription factors regulating gene programs 
  ### ChromVar_Snakemake
  ### get_motif_enrichment

## Testing disease genes for preferential expression in particular cell types
  ### Preferential expression of CHD genes

## Assessing cell type heritability enrichment with stratified LD score regression
  ### LDSC_Snakemake 

## Preparing variants for overlapping with scE2G-predicted enhancers
  ### LD_clumping_expansion

## Linking GWAS variants to enhancers, target genes, cell types, and gene programs
  ### V2G_v2

## Assessing enrichment of genes associated with valve traits in VIC program 27 
  ### Assessing enrichment of genes associated with valve traits in VIC program 27






