# Analysis code for *Molecular convergence of risk variants for congenital heart defects leveraging a regulatory map of the human fetal heart*
All the snakemake pipelines were designed to be run on an HPC platform, [Sherlock](https://www.sherlock.stanford.edu) specifically. The config files are located in \[dir\]/config
## envrionments
  ### environment for snakemake runs:
  envs/snakemake.yml 
  ### enivronemnt for running R scripts:
  envs/R.yml

## scRNA-seq alignment and background removal and Sample demultiplexing
  ### RNA_alignment_snakemake
  This directory contains code for aligning FASTQ files of the RNA portion of multiomic sequencing to a reference assembly using [Starsolo](https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md) and demultiplexing samples using [Souporcell](https://github.com/wheaton5/souporcell). The input sample table requires information about the location of the FASTQ files, the name of the library as appeared in the FASTQ filenames, and the number of samples mixed in the library. 
  ### CellBender_Snakemake 
  This directory contains code for removing ambient RNA (background) using CellBender. 
  
## scATAC-seq alignment
  ### scATAC_pipeline_lite
  This directory contains code for aligning FASTQ files of the ATAC portion of multiomic sequencing to a reference assembly. The code base was forked from https://github.com/austintwang/scATAC_pipeline_lite (commit: 76a4bc545318b5685eb9843e94000dde0d3394e7) with modifications for file and result organization. 

## scATAC-seq quality control
  ### multiomic_ATAC_QC_Snakemake
  The directory contains code for quality control the scATAC-seq portion of the multiome libraries. As shown in the config file, the default filters are minTSS >= 6, and minFrags >= 1000, unless otherwise specified in the sample sheet. Only cells considered to be valid by Starsolo were considered. Both "addDoubletScores" from [ArchR](https://www.archrproject.com/bookdown/inferring-scatac-seq-doublets-with-archr.html) and [Amulet](https://github.com/UcarLab/AMULET) were used to identify doublets. 

## scRNA-seq quality control and normalization
  ### RNA_QC_Snakemake
  The directory contains code for generating the necessary QC metrics of scRNA-seq using the outputs from [ambient RNA removal](#cellbender_snakemake). 
  ### Normalization_Snakemake
  The directory contains code for filtering cells based on the thresholds set by manually examing the QC metrics generated by the previous step. The filtered scRNA count matrices are then normalized and [Scublet](https://github.com/swolock/scrublet) is used to identify doublets. 
  
## Doublet removal
  ### Doublet_removal_Snakemake
  This directory contains code for rerunning Scrublet if specified in the SampleSheet ("Doublet_threshold" != "Default"), removing the union of doublets identified by Scrublet, [Archr](#multiomic_atac_qc_Snakemake), [Amulet](#multiomic_atac_qc_Snakemake), and [Souporcell])(#rna_alignment_snakemake). 
## Clustering and cell type annotation
  ### evaluation_Snakemake
  This directory contain code for merging filtered Seurat object, normalizing the resulting count matrix and clustering at various population. It also generates heatmaps that reflect the marker gene expressions in the clusters and tentatively labeling clusters by mapping the clusters to previous annotations. 
  ### clustering_Snakemake
  This directory contains code for performing clustering on each major cell type group. The code performs count spliting, normalizing and clustering of the training dataset at various resolutions, and label the test dataset with the clustering of the training set and highlighting the expression of the marker genes in the clusters in the test set. 
  ### annotation_Snakemake
  This directory contains code for faciliating the annotation of fine-grained clusters. It generates heatmap showing the top 10 differentially expressed genes in the clusters at the resolution specificed in the sample sheet. It also generates the corresponding dot plots and feature plots of specified marker genes. It also calculates the TPM of genes in each cluster. 

## Defining gene programs with consensus non-negative matrix factorization
  ### cNMF_run_and_analysis/Snakefile_cNMF
  This directory contains code for running [cNMF](https://github.com/dylkot/cNMF) on count matrices specified in SampleSheet using an array of Ks specified in the config file. 
  ### cNMF_run_and_analysis/Snakefile_cNMF_analysis
  This directory contains code for performing further analysis on cNMF results. The code reruns the consensus step using an updated local density filter, performs file organization, summarize how gene programs are distributed across various features (cell annotation, sex, etc.), identifies GO terms associated with the gene program, calculates the variants in gene expression explained by the gene programs, and generate additional plots for K selections.

## Identifying transcription factors regulating gene programs 
  ### ChromVar_Snakemake
  This directory contains code for performing [chromVar](https://github.com/GreenleafLab/chromVAR) to calculate per-cell motif accessibility and output the Z-scores the [bias corrected deviations](https://greenleaflab.github.io/chromVAR/articles/Articles/Deviations.html) of each motif in each cell.
  ### get_motif_enrichment
  This directory contains code for calucalting the correaltion bewteen the per-cell motif accessibility and the expression of gene programs. 

## Assessing cell type heritability enrichment with stratified LD score regression
  ### LDSC_Snakemake 
  This directory contains analysis code for LD score regression.
## Preparing variants for overlapping with scE2G-predicted enhancers
  ### LD_clumping_expansion
  This directory contains code for calculating beta and odds ratios of each variants from each dataset, and performs LD clumping and expansion to get the variants at teach locus. 
## Linking GWAS variants to enhancers, target genes, cell types, and gene programs
  ### V2G_v2
  This directory contains code for linking GWAS variants to cell-type specific enhancers, target genes and gene programs. 

## Testing disease genes for preferential expression in particular cell types
  ### CHD_enrichment
  Enrichment of CHD genes in each cell type. 
  
## Assessing enrichment of genes associated with valve traits in VIC program 27 
  ### VIC_27_enrichment
  Enrichment of genes targted by variants linked to valve traits in VIC program 27





